Task 3:- Experience Analytics

Focusing on network parameters like TCP retransmission, Round Trip Time (RTT), Throughput, and the customers’ device characteristics like the handset type to conduct a deep user experience analysis. 

Import Necesarry Packages

In [14]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [15]:
# Read the data

df1 = pd.read_excel('telcom_data.xlsx')

In [16]:
df1

Bearer Id               Start  Start ms  \
0       13114483460844900352 2019-04-04 12:01:18     770.0   
1       13114483482878900224 2019-04-09 13:04:04     235.0   
2       13114483484080500736 2019-04-09 17:42:11       1.0   
3       13114483485442799616 2019-04-10 00:31:25     486.0   
4       13114483499480700928 2019-04-12 20:10:23     565.0   
...                      ...                 ...       ...   
149996   7277825670196679680 2019-04-29 07:28:42     451.0   
149997   7349883264234609664 2019-04-29 07:28:42     483.0   
149998  13114483573367300096 2019-04-29 07:28:43     283.0   
149999  13114483573367300096 2019-04-29 07:28:43     696.0   
150000                                       NaT       NaN   

                       End  End ms  Dur. (ms)          IMSI  MSISDN/Number  \
0      2019-04-25 14:35:31   662.0  1823652.0  2.082014e+14   3.366496e+10   
1      2019-04-25 08:15:48   606.0  1365104.0  2.082019e+14   3.368185e+10   
2      2019-04-25 11:58:13   652.0  1361762.0  2.082003e+14   3.376063e+10   
3      2019-04-25 07:36:35   171.0  1321509.0  2.082014e+14   3.375034e+10   
4      2019-04-25 10:40:32   954.0  1089009.0  2.082014e+14   3.369980e+10   
...                    ...     ...        ...           ...            ...   
149996 2019-04-30 06:02:33   214.0    81230.0  2.082022e+14   3.365069e+10   
149997 2019-04-30 10:41:33   187.0    97970.0  2.082019e+14   3.366345e+10   
149998 2019-04-30 10:46:12   810.0    98249.0  2.082017e+14   3.362189e+10   
149999 2019-04-30 10:40:34   327.0    97910.0  2.082021e+14   3.361962e+10   
150000                 NaT     NaN        NaN           NaN            NaN   

                IMEI Last Location Name  ...  Youtube DL (Bytes)  \
0       3.552121e+13   9164566995485190  ...        1.585461e+07   
1       3.579401e+13            L77566A  ...        2.024740e+07   
2       3.528151e+13            D42335A  ...        1.972566e+07   
3       3.535661e+13            T21824A  ...        2.138812e+07   
4       3.540701e+13            D88865A  ...        1.525938e+07   
...              ...                ...  ...                 ...   
149996  3.548311e+13            D20434A  ...        1.619167e+07   
149997  3.566051e+13            D10223C  ...        1.387723e+07   
149998  3.572121e+13            T51102A  ...        2.266051e+07   
149999  8.618620e+13            L88342B  ...        8.817106e+06   
150000           NaN                NaN  ...        1.163407e+07   

        Youtube UL (Bytes)  Netflix DL (Bytes)  Netflix UL (Bytes)  \
0             2.501332e+06        8.198936e+06        9.656251e+06   
1             1.911173e+07        1.833841e+07        1.722713e+07   
2             1.469958e+07        1.758779e+07        6.163408e+06   
3             1.514664e+07        1.399465e+07        1.097942e+06   
4             1.896287e+07        1.712458e+07        4.152180e+05   
...                    ...                 ...                 ...   
149996        1.176343e+07        1.788370e+07        1.967816e+07   
149997        8.288284e+06        1.935015e+07        2.129315e+07   
149998        1.855903e+06        9.963942e+06        5.065760e+06   
149999        8.305402e+06        3.322253e+06        1.317259e+07   
150000        1.100941e+07        1.162685e+07        1.100175e+07   

        Gaming DL (Bytes)  Gaming UL (Bytes)  Other DL (Bytes)  \
0            2.780823e+08       1.434415e+07      1.717444e+08   
1            6.087501e+08       1.170709e+06      5.269042e+08   
2            2.295846e+08       3.956300e+05      4.106926e+08   
3            7.995382e+08       1.084972e+07      7.490399e+08   
4            5.277072e+08       3.529801e+06      5.507095e+08   
...                   ...                ...               ...   
149996       5.266097e+08       9.197207e+06      3.264510e+06   
149997       6.268931e+08       4.735033e+06      7.121804e+08   
149998       5.535395e+08       1.339432e+07      1.211009e+08   
149999       3.5253

In [17]:
# Display all column names to identify those related to TCP Retransmission, RTT, and Throughput
df1.columns.tolist()


['Bearer Id',
 'Start',
 'Start ms',
 'End',
 'End ms',
 'Dur. (ms)',
 'IMSI',
 'MSISDN/Number',
 'IMEI',
 'Last Location Name',
 'Avg RTT DL (ms)',
 'Avg RTT UL (ms)',
 'Avg Bearer TP DL (kbps)',
 'Avg Bearer TP UL (kbps)',
 'TCP DL Retrans. Vol (Bytes)',
 'TCP UL Retrans. Vol (Bytes)',
 'DL TP < 50 Kbps (%)',
 '50 Kbps < DL TP < 250 Kbps (%)',
 '250 Kbps < DL TP < 1 Mbps (%)',
 'DL TP > 1 Mbps (%)',
 'UL TP < 10 Kbps (%)',
 '10 Kbps < UL TP < 50 Kbps (%)',
 '50 Kbps < UL TP < 300 Kbps (%)',
 'UL TP > 300 Kbps (%)',
 'HTTP DL (Bytes)',
 'HTTP UL (Bytes)',
 'Activity Duration DL (ms)',
 'Activity Duration UL (ms)',
 'Dur. (ms).1',
 'Handset Manufacturer',
 'Handset Type',
 'Nb of sec with 125000B < Vol DL',
 'Nb of sec with 1250B < Vol UL < 6250B',
 'Nb of sec with 31250B < Vol DL < 125000B',
 'Nb of sec with 37500B < Vol UL',
 'Nb of sec with 6250B < Vol DL < 31250B',
 'Nb of sec with 6250B < Vol UL < 37500B',
 'Nb of sec with Vol DL < 6250B',
 'Nb of sec with Vol UL < 1250B',
 'Socia

In [18]:
# Define columns for calculation
tcp_columns = ['TCP DL Retrans. Vol (Bytes)', 'TCP UL Retrans. Vol (Bytes)']
rtt_columns = ['Avg RTT DL (ms)', 'Avg RTT UL (ms)']
throughput_columns = ['Avg Bearer TP DL (kbps)', 'Avg Bearer TP UL (kbps)']

# Handle missing values by replacing them with the mean for numerical columns
for col in tcp_columns + rtt_columns + throughput_columns:
    df1[col].fillna(df1[col].mean(), inplace=True)

# For the Handset Type, replace missing values with the mode
df1['Handset Type'].fillna(df1['Handset Type'].mode()[0], inplace=True)

# Check if there are still any missing values in these columns
df1[tcp_columns + rtt_columns + throughput_columns + ['Handset Type']].isnull().sum()


TCP DL Retrans. Vol (Bytes)    0
TCP UL Retrans. Vol (Bytes)    0
Avg RTT DL (ms)                0
Avg RTT UL (ms)                0
Avg Bearer TP DL (kbps)        0
Avg Bearer TP UL (kbps)        0
Handset Type                   0
dtype: int64

All missing values in the specified columns have been successfully handled. There are no missing values left in the TCP retransmission, RTT, throughput, and handset type columns.

In [19]:
# Outliers Detection and Treatment.

# Let's start by identifying and treating outliers in these columns.

for col in tcp_columns + rtt_columns + throughput_columns:
    Q1 = df1[col].quantile(0.25)
    Q3 = df1[col].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    mean = df1[col].mean()
    outliers = ~df1[col].between(lower_bound, upper_bound)

    # Replace outliers with the mean using .loc
    df1.loc[outliers, col] = mean  

In [20]:
df1

Bearer Id               Start  Start ms  \
0       13114483460844900352 2019-04-04 12:01:18     770.0   
1       13114483482878900224 2019-04-09 13:04:04     235.0   
2       13114483484080500736 2019-04-09 17:42:11       1.0   
3       13114483485442799616 2019-04-10 00:31:25     486.0   
4       13114483499480700928 2019-04-12 20:10:23     565.0   
...                      ...                 ...       ...   
149996   7277825670196679680 2019-04-29 07:28:42     451.0   
149997   7349883264234609664 2019-04-29 07:28:42     483.0   
149998  13114483573367300096 2019-04-29 07:28:43     283.0   
149999  13114483573367300096 2019-04-29 07:28:43     696.0   
150000                                       NaT       NaN   

                       End  End ms  Dur. (ms)          IMSI  MSISDN/Number  \
0      2019-04-25 14:35:31   662.0  1823652.0  2.082014e+14   3.366496e+10   
1      2019-04-25 08:15:48   606.0  1365104.0  2.082019e+14   3.368185e+10   
2      2019-04-25 11:58:13   652.0  1361762.0  2.082003e+14   3.376063e+10   
3      2019-04-25 07:36:35   171.0  1321509.0  2.082014e+14   3.375034e+10   
4      2019-04-25 10:40:32   954.0  1089009.0  2.082014e+14   3.369980e+10   
...                    ...     ...        ...           ...            ...   
149996 2019-04-30 06:02:33   214.0    81230.0  2.082022e+14   3.365069e+10   
149997 2019-04-30 10:41:33   187.0    97970.0  2.082019e+14   3.366345e+10   
149998 2019-04-30 10:46:12   810.0    98249.0  2.082017e+14   3.362189e+10   
149999 2019-04-30 10:40:34   327.0    97910.0  2.082021e+14   3.361962e+10   
150000                 NaT     NaN        NaN           NaN            NaN   

                IMEI Last Location Name  ...  Youtube DL (Bytes)  \
0       3.552121e+13   9164566995485190  ...        1.585461e+07   
1       3.579401e+13            L77566A  ...        2.024740e+07   
2       3.528151e+13            D42335A  ...        1.972566e+07   
3       3.535661e+13            T21824A  ...        2.138812e+07   
4       3.540701e+13            D88865A  ...        1.525938e+07   
...              ...                ...  ...                 ...   
149996  3.548311e+13            D20434A  ...        1.619167e+07   
149997  3.566051e+13            D10223C  ...        1.387723e+07   
149998  3.572121e+13            T51102A  ...        2.266051e+07   
149999  8.618620e+13            L88342B  ...        8.817106e+06   
150000           NaN                NaN  ...        1.163407e+07   

        Youtube UL (Bytes)  Netflix DL (Bytes)  Netflix UL (Bytes)  \
0             2.501332e+06        8.198936e+06        9.656251e+06   
1             1.911173e+07        1.833841e+07        1.722713e+07   
2             1.469958e+07        1.758779e+07        6.163408e+06   
3             1.514664e+07        1.399465e+07        1.097942e+06   
4             1.896287e+07        1.712458e+07        4.152180e+05   
...                    ...                 ...                 ...   
149996        1.176343e+07        1.788370e+07        1.967816e+07   
149997        8.288284e+06        1.935015e+07        2.129315e+07   
149998        1.855903e+06        9.963942e+06        5.065760e+06   
149999        8.305402e+06        3.322253e+06        1.317259e+07   
150000        1.100941e+07        1.162685e+07        1.100175e+07   

        Gaming DL (Bytes)  Gaming UL (Bytes)  Other DL (Bytes)  \
0            2.780823e+08       1.434415e+07      1.717444e+08   
1            6.087501e+08       1.170709e+06      5.269042e+08   
2            2.295846e+08       3.956300e+05      4.106926e+08   
3            7.995382e+08       1.084972e+07      7.490399e+08   
4            5.277072e+08       3.529801e+06      5.507095e+08   
...                   ...                ...               ...   
149996       5.266097e+08       9.197207e+06      3.264510e+06   
149997       6.268931e+08       4.735033e+06      7.121804e+08   
149998       5.535395e+08       1.339432e+07      1.211009e+08   
149999       3.5253

In [21]:
# Calculate the aggregates after outlier treatment

# TCP Retransmission, RTT, and Throughput with outliers treated
df1['Total TCP Retransmission (Bytes)'] = df1[tcp_columns].sum(axis=1)
df1['Avg RTT (ms)'] = df1[rtt_columns].mean(axis=1)
df1['Avg Throughput (kbps)'] = df1[throughput_columns].mean(axis=1)

# Calculate the aggregates
tcp_retransmission_avg = df1.groupby('MSISDN/Number')['Total TCP Retransmission (Bytes)'].mean()
rtt_avg = df1.groupby('MSISDN/Number')['Avg RTT (ms)'].mean()
throughput_avg = df1.groupby('MSISDN/Number')['Avg Throughput (kbps)'].mean()
handset_type_mode = df1.groupby('MSISDN/Number')['Handset Type'].agg(lambda x: x.mode()[0])

# Aggregate data with outliers treated
aggregated_data_outliers_treated = pd.DataFrame({
    'Average TCP Retransmission (Bytes)': tcp_retransmission_avg,
    'Average RTT (ms)': rtt_avg,
    'Average Throughput (kbps)': throughput_avg,
    'Most Common Handset Type': handset_type_mode
})

# Display the first few rows of the aggregated data after outlier treatment
aggregated_data_outliers_treated.head()

Average TCP Retransmission (Bytes)  Average RTT (ms)  \
MSISDN/Number                                                         
3.360100e+10                         2.156957e+07         23.000000   
3.360100e+10                         2.156957e+07         15.500000   
3.360100e+10                         2.156957e+07         63.729294   
3.360101e+10                         7.607247e+05         42.000000   
3.360101e+10                         1.547020e+07         29.750000   

               Average Throughput (kbps)        Most Common Handset Type  
MSISDN/Number                                                             
3.360100e+10                   38.000000  Huawei P20 Lite Huawei Nova 3E  
3.360100e+10                   49.500000          Apple iPhone 7 (A1778)  
3.360100e+10                   48.500000                       undefined  
3.360101e+10                  124.000000         Apple iPhone 5S (A1457)  
3.360101e+10                10551.357162         Apple iPhone Se (A1723)

In [22]:
# Compute & list 10 of the top, bottom, and most frequent:
# a.	TCP values in the dataset. 
# b.	RTT values in the dataset.
# c.	Throughput values in the dataset.

# Reload the dataset to get original values (before outlier treatment)
df_original = pd.read_excel('telcom_data.xlsx')

# Fill missing values with the mean for numerical columns
for col in tcp_columns + rtt_columns + throughput_columns:
    df_original[col].fillna(df_original[col].mean(), inplace=True)

# Compute top 10, bottom 10, and most frequent values for TCP, RTT, and Throughput

# a. TCP values
tcp_total = df_original[tcp_columns].sum(axis=1)
tcp_top10 = tcp_total.nlargest(10)
tcp_bottom10 = tcp_total.nsmallest(10)
tcp_most_frequent = tcp_total.round().value_counts().nlargest(10)

# b. RTT values
rtt_total = df_original[rtt_columns].mean(axis=1)
rtt_top10 = rtt_total.nlargest(10)
rtt_bottom10 = rtt_total.nsmallest(10)
rtt_most_frequent = rtt_total.round().value_counts().nlargest(10)

# c. Throughput values
throughput_total = df_original[throughput_columns].mean(axis=1)
throughput_top10 = throughput_total.nlargest(10)
throughput_bottom10 = throughput_total.nsmallest(10)
throughput_most_frequent = throughput_total.round().value_counts().nlargest(10)

(tcp_top10, tcp_bottom10, tcp_most_frequent, rtt_top10, rtt_bottom10, rtt_most_frequent, throughput_top10, throughput_bottom10, throughput_most_frequent)



(34636     4.344116e+09
 140797    4.301477e+09
 77950     4.294432e+09
 135677    4.291465e+09
 3741      4.288121e+09
 119667    4.275492e+09
 39608     4.268647e+09
 59011     4.261616e+09
 76971     4.260174e+09
 41182     4.254660e+09
 dtype: float64,
 143594     86.0
 111100     97.0
 18417     106.0
 13142     108.0
 97061     113.0
 8794      128.0
 99947     129.0
 83214     134.0
 95955     134.0
 12849     143.0
 dtype: float64,
 21569573.0    85265
 20811208.0      650
 760989.0        249
 20811244.0      248
 759697.0        136
 20811232.0      132
 759751.0        132
 762319.0        121
 760977.0        106
 20812574.0       99
 Name: count, dtype: int64,
 30155     48462.0
 29916     32320.5
 17894     27906.0
 5963      27424.0
 22839     13639.0
 23442     13150.0
 1361      12961.0
 81256     12857.5
 97312     12694.0
 100575    12369.0
 dtype: float64,
 42602     0.0
 103321    0.0
 124538    0.0
 143878    0.0
 71738     1.0
 50966     2.0
 144923    2.0
 10354

TCP Values

Top 10 TCP Values: These are the highest TCP retransmission volumes observed in the dataset, with the highest being around 4,344,116,000 bytes. These values indicate instances where a significant amount of data had to be retransmitted, possibly due to network issues.

Bottom 10 TCP Values: The lowest TCP retransmission volumes, with some values as low as 86 bytes, indicating sessions with minimal retransmission.

Most Frequent TCP Values: The most common TCP retransmission volume is around 21.57 million bytes, occurring in 85265 records. Other frequent values show a wide distribution, indicating varied network conditions among users.

RTT Values

Top 10 RTT Values: The highest RTT values, with the top being 48462 ms, suggest significant delays in data transmission for these instances.

Bottom 10 RTT Values: The lowest RTT values, going down to 0.0 ms

Most Frequent RTT Values: The most common RTT value rounded to the nearest millisecond is 64 ms, found in 28429 records. This range represents typical network latency for a substantial number of users.

Throughput Values

Top 10 Throughput Values: The highest throughput rates observed, with the top at 191131 kbps, showcasing instances of very high data transmission rates.

Bottom 10 Throughput Values: The lowest throughput rates, including several records with 0.0 kbps, indicating no data transmission. Other low values suggest minimal throughput during those sessions.

Most Frequent Throughput Values: The most common throughput value is 48 kbps, occurring in 5067 records. This indicates a relatively low transmission rate as a frequent condition among the dataset records.

In [23]:
# Compute & report:
# d.	The distribution of the average throughput per handset type and provide interpretation for your findings.
# e.	The average TCP retransmission view per handset type and provide interpretation for your findings.

# d. Distribution of the average throughput per handset type
avg_throughput_per_handset = df_original.groupby('Handset Type')[throughput_columns].mean().mean(axis=1)
avg_throughput_per_handset_sorted = avg_throughput_per_handset.sort_values(ascending=False)

# e. The average TCP retransmission view per handset type
tcp_columns_original = ['TCP DL Retrans. Vol (Bytes)', 'TCP UL Retrans. Vol (Bytes)']
avg_tcp_retransmission_per_handset = df_original.groupby('Handset Type')[tcp_columns_original].sum().sum(axis=1)
avg_tcp_retransmission_per_handset_sorted = avg_tcp_retransmission_per_handset.sort_values(ascending=False)

(avg_throughput_per_handset_sorted.head(10), avg_tcp_retransmission_per_handset_sorted.head(10))


(Handset Type
 Huawei B715S-23C               54497.115385
 New-Bund Technol. Mix 2        52605.500000
 Lg G6+                         50067.500000
 Zyxel Communicat. Sbg3600      48675.500000
 Huawei Y9 2019                 45869.500000
 Xiaomi Communica. M1902F1G     42806.833333
 Spa Condor Elect. Allure M2    42155.750000
 Ovvi-Cellphone T. Echo         39476.500000
 Huawei Pele                    39190.500000
 Spa Condor Elect. Plume L2     38264.500000
 dtype: float64,
 Handset Type
 Huawei B528S-23A                1.089757e+12
 Apple iPhone 6S (A1688)         1.717744e+11
 Apple iPhone 6 (A1586)          1.523796e+11
 undefined                       1.472596e+11
 Apple iPhone 7 (A1778)          1.053158e+11
 Apple iPhone Se (A1723)         9.210993e+10
 Apple iPhone 8 (A1905)          8.697379e+10
 Apple iPhone Xr (A2105)         8.410857e+10
 Samsung Galaxy S8 (Sm-G950F)    6.927049e+10
 Huawei E5180                    6.778901e+10
 dtype: float64)

Interpretations :- The average throughput per handset type reveals significant variations in network performance across different devices. The top devices by average throughput include:

Huawei B715S-23C and New-Bund Technol. Mix 2, leading the list with exceptionally high throughput rates. This suggests that these devices, likely due to their advanced hardware and network capabilities, can achieve superior network performance, benefiting from faster data transmission rates.
Devices like Lg G6+, Zyxel Communicat. Sbg3600, and Huawei Y9 2019 also show high throughput, indicating good network utilization.
The presence of various brands and models in the top list suggests that device capabilities, including support for advanced network technologies (e.g., LTE, 5G), play a crucial role in achieving higher throughput.

The average TCP retransmission volume per handset type highlights differences in how devices handle network stability and data transmission:

Huawei B528S-23A shows the highest TCP retransmission volume, followed by models like Apple iPhone 6S (A1688) and Apple iPhone 6 (A1586). High retransmission volumes can indicate issues with network connectivity or stability, leading to more packets being resent.
The appearance of undefined handset types with high retransmission volumes suggests that data for some devices might not be accurately captured, affecting analysis accuracy.
Popular smartphones like Apple iPhone 7 (A1778) and Apple iPhone Se (A1723) and Samsung Galaxy S8 (Sm-G950F) also appear on the list, indicating that while these devices generally offer good performance, they are not immune to network conditions that necessitate packet retransmissions.

In [24]:
# Using the experience metrics above, 
# Perform a k-means clustering (where k = 3) to segment users into groups of experiences and provide a brief description of each cluster. 


from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler

# Selecting the experience metrics for clustering
experience_metrics = df1[['Avg RTT DL (ms)', 'Avg RTT UL (ms)', 
                                  'Avg Bearer TP DL (kbps)', 'Avg Bearer TP UL (kbps)', 
                                  'TCP DL Retrans. Vol (Bytes)', 'TCP UL Retrans. Vol (Bytes)']]

# Standardizing the data
scaler = StandardScaler()
experience_metrics_scaled = scaler.fit_transform(experience_metrics)

# Perform k-means clustering
kmeans = KMeans(n_clusters=3, random_state=42)
clusters = kmeans.fit_predict(experience_metrics_scaled)

# Add the cluster labels to the dataframe
df1['Cluster'] = clusters

# Analyzing the clusters
cluster_summary = df1.groupby('Cluster')[['Avg RTT DL (ms)', 'Avg RTT UL (ms)', 
                                                  'Avg Bearer TP DL (kbps)', 'Avg Bearer TP UL (kbps)', 
                                                  'TCP DL Retrans. Vol (Bytes)', 'TCP UL Retrans. Vol (Bytes)']].mean()

cluster_summary


Avg RTT DL (ms)  Avg RTT UL (ms)  Avg Bearer TP DL (kbps)  \
Cluster                                                              
0              63.260866         8.720111               476.964401   
1              87.351248        22.912181             25408.821058   
2              63.907514         7.973845             16717.325692   

         Avg Bearer TP UL (kbps)  TCP DL Retrans. Vol (Bytes)  \
Cluster                                                         
0                      91.813861                 1.974941e+07   
1                    1630.014872                 8.176710e+06   
2                    1162.309155                 1.810936e+06   

         TCP UL Retrans. Vol (Bytes)  
Cluster                               
0                      743255.873406  
1                      204054.767701  
2                      158249.848737

Interpretation :- 

By segmenting users based on these metrics, we can identify specific groups that may benefit from targeted improvements in network infrastructure, optimization of network settings, or customer support interventions. 

For instance:

Users in the High Performance cluster might be leveraged as benchmarks for network capabilities.
Strategies to elevate users from the Moderate to High Performance cluster could focus on reducing RTT and TCP retransmissions.
For those in the Low Performance cluster, investigating and addressing the root causes of poor performance becomes critical, possibly including network upgrades or advising on device replacements.

In [25]:
df1

Bearer Id               Start  Start ms  \
0       13114483460844900352 2019-04-04 12:01:18     770.0   
1       13114483482878900224 2019-04-09 13:04:04     235.0   
2       13114483484080500736 2019-04-09 17:42:11       1.0   
3       13114483485442799616 2019-04-10 00:31:25     486.0   
4       13114483499480700928 2019-04-12 20:10:23     565.0   
...                      ...                 ...       ...   
149996   7277825670196679680 2019-04-29 07:28:42     451.0   
149997   7349883264234609664 2019-04-29 07:28:42     483.0   
149998  13114483573367300096 2019-04-29 07:28:43     283.0   
149999  13114483573367300096 2019-04-29 07:28:43     696.0   
150000                                       NaT       NaN   

                       End  End ms  Dur. (ms)          IMSI  MSISDN/Number  \
0      2019-04-25 14:35:31   662.0  1823652.0  2.082014e+14   3.366496e+10   
1      2019-04-25 08:15:48   606.0  1365104.0  2.082019e+14   3.368185e+10   
2      2019-04-25 11:58:13   652.0  1361762.0  2.082003e+14   3.376063e+10   
3      2019-04-25 07:36:35   171.0  1321509.0  2.082014e+14   3.375034e+10   
4      2019-04-25 10:40:32   954.0  1089009.0  2.082014e+14   3.369980e+10   
...                    ...     ...        ...           ...            ...   
149996 2019-04-30 06:02:33   214.0    81230.0  2.082022e+14   3.365069e+10   
149997 2019-04-30 10:41:33   187.0    97970.0  2.082019e+14   3.366345e+10   
149998 2019-04-30 10:46:12   810.0    98249.0  2.082017e+14   3.362189e+10   
149999 2019-04-30 10:40:34   327.0    97910.0  2.082021e+14   3.361962e+10   
150000                 NaT     NaN        NaN           NaN            NaN   

                IMEI Last Location Name  ...  Gaming DL (Bytes)  \
0       3.552121e+13   9164566995485190  ...       2.780823e+08   
1       3.579401e+13            L77566A  ...       6.087501e+08   
2       3.528151e+13            D42335A  ...       2.295846e+08   
3       3.535661e+13            T21824A  ...       7.995382e+08   
4       3.540701e+13            D88865A  ...       5.277072e+08   
...              ...                ...  ...                ...   
149996  3.548311e+13            D20434A  ...       5.266097e+08   
149997  3.566051e+13            D10223C  ...       6.268931e+08   
149998  3.572121e+13            T51102A  ...       5.535395e+08   
149999  8.618620e+13            L88342B  ...       3.525370e+08   
150000           NaN                NaN  ...       4.220447e+08   

        Gaming UL (Bytes)  Other DL (Bytes)  Other UL (Bytes)  \
0            1.434415e+07      1.717444e+08      8.814393e+06   
1            1.170709e+06      5.269042e+08      1.505514e+07   
2            3.956300e+05      4.106926e+08      4.215763e+06   
3            1.084972e+07      7.490399e+08      1.279728e+07   
4            3.529801e+06      5.507095e+08      1.391032e+07   
...                   ...               ...               ...   
149996       9.197207e+06      3.264510e+06      1.348742e+07   
149997       4.735033e+06      7.121804e+08      2.457758e+06   
149998       1.339432e+07      1.211009e+08      1.131473e+07   
149999       2.529475e+06      8.147131e+08      1.406930e+06   
150000       8.288398e+06      4.211005e+08      8.264799e+06   

        Total UL (Bytes)  Total DL (Bytes)  Total TCP Retransmission (Bytes)  \
0             36749741.0       308879636.0                      2.156957e+07   
1             53800391.0       653384965.0                      2.156957e+07   
2             27883638.0       279807335.0                      2.156957e+07   
3             43324218.0       846028530.0                      2.156957e+07   
4             38542814.0       569138589.0                      2.156957e+07   
...                  ...               ...                               ...   
149996        57628851.0       574175259.0                      2.156957e+07   
149997        39135081.0       666648844.0                      2.156957e+07   
149998        34912224.0       

In [26]:
df1.to_excel('Experience Analytics.xlsx')